File: old/2017/python/01.py
# Advent of Code 2017, day 1
# (c) aichingert

def part1(inp):
    ans = 0
    if inp[0] == inp[-1]: ans = ord(inp[0])-ord('0')

    for i in range(len(inp)-1):
        if inp[i] == inp[i+1]:
            ans += ord(inp[i]) - ord('0')
    return ans
def part2(inp):
    ans = 0
    l = len(inp)//2
    for i in range(len(inp)):
        if inp[i] == inp[(l+i)%len(inp)]:
            ans += ord(inp[i]) - ord('0')
    return ans

inp=open("../input/01").read().strip()

print("Part 1:", part1(inp))
print("Part 2:", part2(inp))

File: old/2017/rust/01.rs
// Advent of Code 2017, day 1
// (c) aichingert

fn part1(inp: &[char]) -> u32 {
    let ans = if inp[0] == inp[inp.len()-1] { ((inp[0] as u8) - ('0' as u8)) as u32 } else { 0 };
    (0..inp.len()-1).map(|i| if inp[i] == inp[i+1] { ((inp[i] as u8) - ('0' as u8)) as u32 } else { 0 } ).sum::<u32>() + ans
}

fn part2(inp: &[char]) -> u32 {
    let half: usize = inp.len()/2;
    (0..inp.len()).map(|i| if inp[i] == inp[(half+i)%inp.len()] { ((inp[i] as u8) - ('0' as u8)) as u32 } else { 0 } ).sum::<u32>()
}

fn main() {
    let inp = std::fs::read_to_string("../input/01").unwrap();
    let inp = inp.trim().chars().collect::<Vec<char>>();

    println!("Part 1: {}", part1(&inp));
    println!("Part 2: {}", part2(&inp));
}
